cmd/jsontoml/main.go

```diff
@@ -45,6 +45,7 @@ func convert(r io.Reader, w io.Writer) error {
 	var v interface{}
 
 	d := json.NewDecoder(r)
+	d.UseNumber()
```

Comment:
this can be questionable as a backward incompatible change.
 
I don't mind removing `d.UseNumber()` from jsontoml file, what I am truly interested in is `json.Number` encoding support.

Reply:
I think we can put that behind a flag for the command line argument.

Interestingly, my main concern about this PR at the moment is adding the rules for encoding `json.Number` directly into the marshaler, which would be a change in behavior for others. I'm a little torn about this at the moment. If the use case is just for the `jsontoml` cli, we can have it make a pass to transform the `json.Number` values in `v` to integers. Would doing a similar pre-pass be an option in your usage of the lib?

Reply:
> I think we can put that behind a flag for the command line argument.

I glanced on how to do that but failed, seems jsontoml and tomljson uses an _internal cli_ package. I can take another look tho or just remove it from this PR and submit another one when ready.

> my main concern about this PR at the moment is adding the rules for encoding json.Number directly into the marshaler, which would be a change in behavior for others.

Oh, the problem is that json.Number implements the TextMarshaller interface but it will serialize numbers as TOML text, not floats nor integers. I tend to think this is a bug and json.Number marshalling is broken for everyone at the moment.

I was hoping this change enters the *features* category about matching `encoding/json` behavior mentioned at https://github.com/pelletier/go-toml?tab=readme-ov-file#stdlib-behavior:

> ## Stdlib behavior
> As much as possible, this library is designed to behave similarly as the standard library's encoding/json.

Reply:
> I glanced on how to do that but failed, seems jsontoml and tomljson uses an internal cli package. I can take another look tho or just remove it from this PR and submit another one when ready.

You should be able to add a new argument to `flag` before `p.Execute()`, and the the value of that flag in the `convert` function.

> Oh, the problem is that json.Number implements the TextMarshaller interface but it will serialize numbers as TOML text, not floats nor integers. I tend to think this is a bug and json.Number marshalling is broken for everyone at the moment.
>
> I was hoping this change enters the features category about matching encoding/json behavior

This is a fair point, that's why I'm considering it, not saying no. :) My take is that ideally, this feature should be achieved by bringing back the `toml.Marshaler` interface. But that's a bit more of an undertaking, and I don't want to block you on it. In the meantime, would you be open to put the marshaling logic behind a flag of a new `unstable.Decoder`? (same request as https://github.com/pelletier/go-toml/pull/921) That way it does not impact the public API and default behavior, and you should be unblocked far more quickly.

Reply:
I keep thinking on alternatives like that prevent double iterating the struct before calling the marshaller:
1. Add a `SetJsonNumber(bool)` method to the encoder <-- the simpler 
2. Add a `SetMarshalJSON(bool)` <-- supports any structs that already implement it 
3. Add a `SetCustomEncoder(type any, fn EncoderFunc)` <-- the most generic, not sure it is worth.

Reply:
I think eventually 3 could be useful, but I'd want to make sure the performance impact is controlled. In the meantime I think `SetMarshalJsonNumbers(bool)` is the most straightforward and the least confusing. Accepting any `json.Marshaler` object feels weird, since that method returns JSON bytes, which then would need to be decoded, and then re-encoded.

Reply:
@pelletier I opted for option (1) in https://github.com/pelletier/go-toml/pull/923/commits/c447b2f6cea2fa076b1ca3951cce346a1bad9472

but can revert or change to whatever suits better it to make this happen 🙏

Reply:
> I think `SetMarshalJsonNumbers(bool)` is the most straightforward and the least confusing.

Cool, renamed in https://github.com/pelletier/go-toml/pull/923/commits/7b5aa9f10da727d9507cdaf2662cd7b41b700cc3